### PR TITLE
Fix/write s3 organized

### PR DESCRIPTION
**Pull request by bogdanguranda**
Done, please check this out :smiley:

### CODE REVIEW

File: reader.go [hunk 1]
```diff
@@ -126,3 +128,46 @@ func (sfr *FactsetReader) unzip(archive string, name string, dest string) error
...
+
+	foundMatches := regex.FindStringSubmatch(filename)
+	if foundMatches == nil {
+		return "", errors.New("The full version is missing or not correctly specified! Filename was " + filename)
+	}
+	if len(foundMatches) > 1 {
+		return "", errors.New("More than 1 full version found!")
+	}
+
+	versionWithExt := regex.FindStringSubmatch(filename)[0]
```
Comment (MMoisa) [inline reader.go]: Why do you call FindStringSubmatch again when you already have foundMatches?

File: reader.go [hunk 2]
```diff
@@ -67,34 +68,35 @@ func (sfr *FactsetReader) download(filePath string, fileName string, dest string
...
+
+		if (majorVersion > foundFile.majorVersion) ||
+			(majorVersion == foundFile.majorVersion && minorVersion > foundFile.minorVersion) {
+			foundFile.name = name
+			foundFile.majorVersion = majorVersion
+			foundFile.minorVersion = minorVersion
 		}
 	}
-	return recFile.name, nil
+	return foundFile.name, nil
```
Comment (kovacshuni) [inline reader.go:105]: There should be a check if the file was found at all, if it's a valid case. Also on the err!=nil to log something if it's handy, and not spammy.

**Review by MMoisa — CHANGES_REQUESTED** [review]

File: reader.go (hunk 2 above)
Comment (bogdanguranda) [reply reader.go:105]: The case is valid, so I treat it with returning an error, as all errors are logged at a higher level. So there will be a log entry for this. Done.

File: reader.go (hunk 1 above)
Comment (bogdanguranda) [reply reader.go]: You are right, there is no reason, thanks for spotting this. Updated repo.

**Review by MMoisa — APPROVED** [review]

**Review by kovacshuni — APPROVED** [review]

File: reader.go [hunk 3]
```diff
@@ -67,34 +68,40 @@ func (sfr *FactsetReader) download(filePath string, fileName string, dest string
...
 	for _, file := range files {
 		name := file.Name()
-		if !strings.Contains(name, searchedRes) {
+		if !strings.Contains(name, searchedFileName) {
 			continue
 		}
-		s := r.FindStringSubmatch(name)[0]
 
-		v, err := strconv.Atoi(strings.TrimSuffix(s, ".zip"))
+		fullVersion, err := GetFullVersion(name)
```
Comment (danielcopaciu) [inline reader.go]: Has GetFullVersion function to be public?

File: reader.go [hunk 4]
```diff
@@ -126,3 +133,44 @@ func (sfr *FactsetReader) unzip(archive string, name string, dest string) error
 	}
 	return nil
 }
+
+func GetFullVersion(filename string) (string, error) {
```
Comment (danielcopaciu) [inline reader.go]: Please avoid using stateless methods.

File: reader.go (hunk 3 above)
Comment (bogdanguranda) [reply reader.go]: It has to be in the interface, so yes, it has to be public, but not static (stateless), so I changed it to be part of interface.

**Review by danielcopaciu — APPROVED** [review]

**Pull request merged**